Added command to add a meme of user's choosing.

naps250 opened the pull request:
Implemented IEquitable for MemeCommand.
Changed design of CommandNames field.
Various QOL improvements.
Small bug fixes.

naps250 committed:
Added command to add a meme of user's choosing.
Implemented IEquitable for MemeCommand.
Changed design of CommandNames field.
Various QOL improvements.
Small bug fixes.

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -13,20 +13,59 @@ namespace SteakBot.Core.EventHandlers.CustomMessageHandlers.CommandMessageHandle
-        internal static IList<MemeCommand> Commands { get; } = LoadCommands();
+        internal static IList<MemeCommand> Commands { get; set; } = LoadCommands();
+
+        private static IEnumerable<string> commandNames;
```
This seems redundant?

penev92 replied:
Also should be `_commandNames` since it's private.

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -13,20 +13,59 @@ namespace SteakBot.Core.EventHandlers.CustomMessageHandlers.CommandMessageHandle
+
+        private static IEnumerable<string> commandNames;
+
+        protected override IEnumerable<string> CommandNames { get => commandNames; set => commandNames = value; }
```
The `This seems redundant?` comment on the private static `commandNames` above comes from here. This seems like it can just be `CommandNames { get; set; }` ?

penev92 replied:
At the end of the day this becomes redundant to the inherited property because it's identical.

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -66,6 +105,22 @@ private static IList<MemeCommand> LoadCommands()
+        private static void ReloadCommands()
+        {
+            Commands = LoadCommands();
+            commandNames = Commands.Select(x => x.Name);
```
This can also change to use the `CommandNames` property directly.

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -66,6 +105,22 @@ private static IList<MemeCommand> LoadCommands()
             }
         }
 
+        private static void ReloadCommands()
```
Using `CommandNames` requires this to become non-static.

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -13,20 +13,59 @@ namespace SteakBot.Core.EventHandlers.CustomMessageHandlers.CommandMessageHandle
+            commandNames = Commands.Select(x => x.Name);
+        }
+
+        public static bool SaveCommand(MemeCommand newCmd)
```
Making `ReloadCommands()` non-static requires this to become non-static as well.

penev92 commented on SteakBot.Core/Modules/AddMemeModule.cs:
```diff
@@ -0,0 +1,69 @@
+{
+    public class AddMemeModule : ModuleBase<SocketCommandContext>
+    {
+        private const char ARG_SEPARATOR = ',';
```
C# conventions for constants are to use PascalCase naming.

penev92 replied:
With the removal of the manual parsing below, this separator becomes redundant.

penev92 commented on SteakBot.Core/Modules/AddMemeModule.cs:
```diff
@@ -0,0 +1,69 @@
+    {
+        private const char ARG_SEPARATOR = ',';
+
+        [Command("add")]
```
I'd like the command to be a bit more descriptive, please.

penev92 commented on SteakBot.Core/Modules/AddMemeModule.cs:
```diff
@@ -0,0 +1,69 @@
+        public async Task AddMeme([Remainder]string message)
+        {
+            bool isMessage = true;
+            string replyMessage = string.Empty;
```
Please use `var` instead of explicit typing wherever possible.

penev92 commented on SteakBot.Core/Modules/AddMemeModule.cs:
```diff
@@ -0,0 +1,69 @@
+        private const char ARG_SEPARATOR = ',';
+
+        [Command("add")]
+        public async Task AddMeme([Remainder]string message)
```
You can let `Discord.Net` handle the parsing (and valitation) for you by changing this to
```cs
        public async Task AddMeme(string name, string value, string description)
```

penev92 replied:
Then the input template will change from
```
!add name,value,description
```
to
```
!add name value description
```
and will even handle multi-word values with quotes: `"This is my name string"`

penev92 commented on SteakBot.Core/Modules/AddMemeModule.cs:
```diff
@@ -0,0 +1,69 @@
+            EmbedBuilder embed = null;
+
+            var args = message.Split(ARG_SEPARATOR).ToList();
+            if (args.Count != 3)
```
With the change of the method's signature and using the built-in argument parsing, this entire `if/else` becomes obsolete.
Now you only need the body of the `else`.

penev92 commented on SteakBot.Core/Modules/AddMemeModule.cs:
```diff
@@ -0,0 +1,69 @@
+            {
+                MemeResultType type = args[1].ToLower().Contains("http") ? MemeResultType.Image : MemeResultType.Text;
+
+                var result = CustomCommandMessageHandler.SaveCommand(new MemeCommand(type, args[0], args[1], args[2]));
```
Since `CustomCommandMessageHandler.SaveCommand()` is no longer static, this will need to change.

penev92 commented on SteakBot.Core/Modules/AddMemeModule.cs:
```diff
@@ -0,0 +1,69 @@
+                  : ReplyAsync(replyMessage, embed:embed.Build()));
+        }
+
+        private IEnumerable<IUser> GetUsers()
```
This method isn't used anywhere. Leftover code?

penev92 commented on SteakBot.Core/Objects/MemeCommand.cs:
```diff
@@ -19,5 +20,63 @@ public MemeCommand(MemeResultType resultType, string name, string value, string
+                return false;
+            }
+
+            return Equals(item);
```
All this can be trimmed down to
```cs
        public override bool Equals(object other)
        {
            var item = other as MemeCommand;
            return item != null && Equals(item);
        }
```

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -13,20 +13,59 @@ namespace SteakBot.Core.EventHandlers.CustomMessageHandlers.CommandMessageHandle
+                    }
+                }
+
+#if DEBUG
```
Do we really need this to stay?

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -13,20 +13,59 @@ namespace SteakBot.Core.EventHandlers.CustomMessageHandlers.CommandMessageHandle
+                }
+
+#if DEBUG
+                // replace original file
```
Comments should start with a capital letter and preferably and with a fullstop or similar.

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/BaseCommandMessageHandler.cs:
```diff
@@ -11,7 +11,7 @@ internal abstract class BaseCommandMessageHandler : ICustomMessageHandler
         protected const string DeleteMessageChar = "!";
 
-        protected IEnumerable<string> CommandNames;
+        protected virtual IEnumerable<string> CommandNames { get; set; }
```
Once the other changes are adressed this can go back to being a protected field or a protected non-virtual property.

naps250 committed:
Bunch of classes now public (were internal).
Changed add command to addMeme and it's not using Discord's parameter validation handling.
Number of small improvements.

naps250 commented:
Addressed all your comments. Please review.

naps250 committed:
Added MemeService - used by AddMemeModule to save and by custom message handler to load commands.

naps250 committed:
Renamed MemeModule to ChatModule.
Renamed AddMemeModule to MemeModule.
Removed ununsed usings.
Moved Commands collection to the MemeService class.
Reverted internal->public access level change of Base and CustomCommandMessageHandler classes.
Moved Audio and MemeService classes to new folder 'Services'.
Various small tweaks to addMeme command in MemeModule.

naps250 committed:
no message

naps250 committed:
Added missing event invoke.

penev92 commented on SteakBot.Core/Objects/MemeCommand.cs:
```diff
@@ -1,8 +1,9 @@
 ﻿using SteakBot.Core.Objects.Enums;
+using System;
```
`using` statement order violation.

penev92 commented on SteakBot.Core/Modules/MemeModule.cs:
```diff
@@ -1,20 +1,55 @@
+                    replyMessage = value;
+                }
+
+                await (isMessage
```
Move this to after the `else` as 
```cs
await ReplyAsync(replyMessage, embed: embed?.Build());
```

penev92 commented on SteakBot.Core/Modules/MemeModule.cs:
```diff
@@ -1,20 +1,55 @@
+
+            var result = _memeService.SaveCommand(new MemeCommand(type, name, value, description));
+
+            var isMessage = type == MemeResultType.Text;
```
With the simplification of the reply this becomes unused.

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,90 @@
+                }
+#endif
+
+                ReloadCommandsEvent.Invoke();
```
Possible NRE. Suggested fix is using null propagation
```cs
ReloadCommandsEvent?.Invoke();
```

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,90 @@
+
+        private bool ValidateNewCommand(MemeCommand newCmd)
+        {
+            if (MemeCommands.Contains(newCmd))
```
How about
```cs
return !MemeCommands.Contains(newCmd);
```

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,90 @@
+
+        public bool SaveCommand(MemeCommand newCmd)
+        {
+            if (ValidateNewCommand(newCmd))
```
Invert `if` statement to reduce nesting.

naps250 committed:
QOL improvements/refactoring.

naps250 commented:
Addressed all your comments (again).

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -1,32 +1,28 @@
             var commandText = message.Content.Replace(CommandChar, "").Replace(DeleteMessageChar, "");
-            var command = Commands.Single(x => x.Name == commandText);
-            switch (command.ResultType)
+            var command = _memeService.MemeCommands.SingleOrDefault(x => x.Name == commandText);
```
The `InvokeInner` is only called once we've verified that the command can be handled by this handler, so there shouldn't be any scenario where this lookup fails.
Can safely retyrn this to `.Single()` and remove the null propagation on the next line.

penev92 commented on SteakBot.Core/EventHandlers/CustomMessageHandlers/CommandMessageHandlers/CustomCommandMessageHandler.cs:
```diff
@@ -54,16 +50,9 @@ protected override bool InvokeInner(SocketUserMessage message)
         #region Private methods
 
-        private static IList<MemeCommand> LoadCommands()
+        public void ReloadCommandNames()
```
This isn't used anywhere outside of this class so can go back to being `private`.

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,85 @@
+        public delegate void ReloadCommands();
+        public event ReloadCommands ReloadCommandsEvent;
+
+        public IList<MemeCommand> MemeCommands { get; set; }
```
`private set;` please.

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,85 @@
+            return true;
+        }
+
+        public IList<MemeCommand> LoadCommands()
```
This should go with the private methods.

penev92 replied:
Return value is never used so you can change to `void` and remove the `return` statement.

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,85 @@
+                }
+            }
+
+#if DEBUG
```
I still don't understand the point of this block of code? Please elaborate.
To me it seems redundant.

naps250 replied:
This was originally intended so as not to lose your settings between debug sessions.
Now I think it can be removed.

penev92 reviewed (changes requested):
I apologize for the long delay. This is turning out nicely and there are only a few small details left to iron out.

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,67 @@
+using SteakBot.Core.Objects;
+using System.Collections.Generic;
+using System.IO;
+using System.Configuration;
```
Using statement order violation.

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,67 @@
+    public class MemeService
+    {
+        private static readonly string MemeCommandsFileName = ConfigurationManager.AppSettings["memeCommandsRelativeFilePath"];
+        private static readonly string MemeCommandsOriginRelativeFilePath = ConfigurationManager.AppSettings["memeCommandsOriginRelativeFilePath"];
```
This is now unused.

penev92 commented on SteakBot/App.config:
```diff
@@ -3,6 +3,7 @@
     <appSettings>
         <add key="BotToken" value=""/>
         <add key="memeCommandsRelativeFilePath" value="Assets/memeCommands.json"/>
+        <add key="memeCommandsOriginRelativeFilePath" value="../../../SteakBot.Core/Assets/memeCommands.json"/>
```
This is now obsolete.

rivanov123 commented on SteakBot.Core/Objects/MemeCommand.cs:
```diff
@@ -19,5 +20,57 @@ public MemeCommand(MemeResultType resultType, string name, string value, string
+                return false;
+            }
+
+            return (object.ReferenceEquals(this.ResultType, other.ResultType)
```
object.ReferenceEquals is always false because it is called on value type. C'MON @naps250

rivanov123 commented on SteakBot.Core/Objects/MemeCommand.cs:
```diff
@@ -19,5 +20,57 @@ public MemeCommand(MemeResultType resultType, string name, string value, string
+
+        public override int GetHashCode()
+        {
+            return this.ResultType.GetHashCode()
```
dont use XOR for combining hashes
Instance A: Prop1 = 1 Prop2 = 0 Instance B: Prop1 = 0 Prop2 = 1 => XOR produces equal hashses.

rivanov123 commented on SteakBot.Core/Objects/MemeCommand.cs:
```diff
@@ -19,5 +20,57 @@ public MemeCommand(MemeResultType resultType, string name, string value, string
+                    || this.Name != null
+                    && this.Name.Equals(other.Name))
+                && (object.ReferenceEquals(this.Value, other.Value)
+                    || this.Value != null
```
Use string == operator. It is overloaded and does the same thing as this.Value != null && this.Value.Equal(other.Value),
null check becomes redundand when using == operator on string.
Equals operator on strings include reference checks, so object.ReferenceEquals on strings is redundand.

rivanov123 commented on SteakBot.Core/Objects/Enums/MemeResultType.cs:
```diff
@@ -1,7 +1,8 @@
 ﻿namespace SteakBot.Core.Objects.Enums
 {
-    internal enum MemeResultType
+    public enum MemeResultType
```
Enums are bad design when used as a branching mechanism, refactor!

penev92 replied:
This is not part of this PR, so feel free to refactor in a follow-up.

rivanov123 commented on SteakBot.Core/Objects/MemeCommand.cs:
```diff
@@ -19,5 +20,57 @@ public MemeCommand(MemeResultType resultType, string name, string value, string
+
+        public bool Equals(MemeCommand other)
+        {
+            if (ReferenceEquals(other, null))
```
Add ReferenceEquals for other and this to check if not passing the same pointer. Currently if other is not equal to null values will be compared directly even if the same pointer as this is passed

penev92 commented on SteakBot.Core/Objects/MemeCommand.cs:
```diff
@@ -19,5 +20,61 @@ public MemeCommand(MemeResultType resultType, string name, string value, string
+            {
+                return false;
+            }
+            if(ReferenceEquals(other, this))
```
Empty line after closing a block, please.

penev92 commented on SteakBot.Core/Services/MemeService.cs:
```diff
@@ -0,0 +1,68 @@
+using Newtonsoft.Json;
+
+
+
```
Only one empty line at a time.

penev92 reviewed (commented):
Just two more small fixes to go.

penev92 reviewed (approved):
Should be good now 👍

penev92 commented:
Sorry this took so long. It is now ready for your next refactor ;)
